satellite/repair/repairer/ec.go

```diff
@@ -0,0 +1,450 @@
+	pieceReaders := make(map[int]io.ReadCloser)
+
+	limiter := sync2.NewLimiter(es.RequiredCount())
+	cond := sync.NewCond(&sync.Mutex{})
+
+	for currentLimitIndex, limit := range limits {
+		if limit == nil {
+			continue
+		}
+
+		currentLimitIndex, limit := currentLimitIndex, limit
+		limiter.Go(ctx, func() {
+			cond.L.Lock()
+			defer cond.Signal()
+			defer cond.L.Unlock()
+
+			for {
+				if successfulPieces >= es.RequiredCount() {
+					// already downloaded minimum number of pieces
+					cond.Broadcast()
+					return
+				}
+				if successfulPieces+inProgress+unusedLimits < es.RequiredCount() {
+					// not enough available limits left to get required number of pieces
+					cond.Broadcast()
+					return
+				}
+
+				if successfulPieces+inProgress >= es.RequiredCount() {
+					cond.Wait()
+					continue
+				}
+
+				unusedLimits--
+				inProgress++
+				cond.L.Unlock()
+
+				downloadedPiece, err := ec.downloadAndVerifyPiece(ctx, limit, privateKey, pieceSize, es.ErasureShareSize())
+				cond.L.Lock()
+				inProgress--
+				if err != nil {
+					ec.log.Debug("Failed to download pieces for repair.", zap.Error(err))
+					return
+				}
+
+				pieceReaders[currentLimitIndex] = ioutil.NopCloser(bytes.NewReader(downloadedPiece))
+				successfulPieces++
+
+				return
+			}
+		})
+	}
+
+	limiter.Wait()
+
+	if successfulPieces < es.RequiredCount() {
+		return nil, Error.New("couldn't download enough pieces, number of successful downloaded pieces (%d) is less than required number (%d)", successfulPieces, es.RequiredCount())
+	}
+
+	fec, err := infectious.NewFEC(es.RequiredCount(), es.TotalCount())
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+
+	esScheme := eestream.NewUnsafeRSScheme(fec, es.ErasureShareSize())
+	expectedSize := pieceSize * int64(es.RequiredCount())
+
+	ctx, cancel := context.WithCancel(ctx)
+	decodeReader := eestream.DecodeReaders(ctx, cancel, ec.log.Named("decode readers"), pieceReaders, esScheme, expectedSize, 0, false)
+
+	return decodeReader, nil
+}
+
+// downloadAndVerifyPiece downloads a piece from a storagenode,
+// expects the original order limit to have the correct piece public key,
+// and expects the hash of the data to match the signed hash provided by the storagenode.
+func (ec *ECRepairer) downloadAndVerifyPiece(ctx context.Context, limit *pb.AddressedOrderLimit, privateKey storj.PiecePrivateKey, pieceSize int64, bufferSize int) (data []byte, err error) {
+	// contact node
+	ps, err := ec.dialPiecestore(ctx, &pb.Node{
+		Id:      limit.GetLimit().StorageNodeId,
+		Address: limit.GetStorageNodeAddress(),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	downloader, err := ps.Download(ctx, limit.GetLimit(), privateKey, 0, pieceSize)
+	if err != nil {
+		return nil, err
+	}
+	defer func() { err = errs.Combine(err, downloader.Close()) }()
+
+	var calculatedHash []byte
+	pieceBytes, err := ioutil.ReadAll(downloader)
+	if err != nil {
+		return nil, err
+	}
+	calculatedHash = pkcrypto.SHA256Hash(pieceBytes)
```

Comment:
this should be moved just before place its used. Calculating hash is CPU consuming and if we can avoid it with simple validation we need to do this.